#!/usr/bin/env Rscript

# plot distances form founder infered by prank to actual founder.
#
# Works of files created by a simulation process to mimic some of
# those that we obtain from the RV217 project.  The RV217 sample ahave
# around 10-12 sequences per sample, all taken from the c2v3c3 region
# gp-120 from the HIV-1 genome.

# install any missing packages
# http://stackoverflow.com/a/19873732/1135316
if (!require("pacman")) install.packages("pacman")
pacman::p_load(ape, dplyr, ggplot2, stringr, Biostrings, readr, tidyr)

## suppressPackageStartupMessages(library(ape))
## suppressPackageStartupMessages(library(dplyr))

## library(Biostrings) 
## library(readr) 
## library(ggplot2) 
## library(stringr) # for str_extract

setwd("~/src/matsen/hiv-sim/sim.rv217/analysis")

files <- list.files(path='../build', pattern="sample\\.fa$", recursive=TRUE)
df <- as_data_frame(list(dir=dirname(files))) %>% 
    mutate(rate=factor(as.numeric(str_match(dir, "^([^/]+)/")[,2]))) %>%
    mutate(rate=reorder(rate, as.numeric(as.character(rate)), FUN=max)) %>%
    mutate(fitness=str_match(dir, "/([^/]+)_[^/]+/replicate")[,2]) %>%
    mutate(fitness=ifelse(is.na(fitness), 'none', fitness)) %>%
    mutate(fitness=factor(fitness)) %>%
    mutate(indel=is.na(str_extract(dir, "noindel"))) %>%
    mutate(replicate=factor(str_match(dir, "replicate_([0-9]*)")[,2])) %>%
    mutate(replicate=reorder(replicate, as.numeric(as.character(replicate)), FUN=max)) %>%
    mutate(wpi=factor(str_match(dir, "/([0-9]+)$")[,2])) %>%
    mutate(wpi=reorder(wpi, as.integer(as.character(wpi)), FUN=max)) %>%
    mutate(dir=file.path('..', 'build', dir))


root.id <- function(file) {
    tree <- ape::read.tree(file, comment.char = "")
    return(tree$node.label[1])
}

# Extract the inferred sequence at the root node of PRANK output
#
# 'dir' parameter is a character string naming a directory where the PRANK
# output files can be found.  The directory is expected to have a file
# 'prank.best.anc.dnd' containing a tree generated by PRANK, and a
# file 'prank.best.anc.fas' containing the sequences at each node of
# the tree in FASTA format.
#
prank.sequence <- function(dir) {
    treefile <- file.path(dir, 'prank.best.anc.dnd')
    rootid <- root.id(treefile)
    fastafile <- file.path(dir, 'prank.best.anc.fas')
    fasta <- read.dna(fastafile, format = 'fasta', comment.char = '')
    return(fasta[which(labels(fasta)==rootid),])
}

founder.sequence <- function() {
    fastafile <- file.path('..', '..', 'templates', 'HIV1C2C3.fasta')
    fasta <- read.dna(fastafile, format = "fasta", comment.char = "")
    return(paste(fasta[1,],collapse=''))
}


beast.distance <- function(dir, model='K80', founder=founder) {
    #print(file)
    burnin_pct = 0.80
    distance <- function(s, model, founder) {
        cat('.')
        alignment <- pwalign(s, founder)
        dist.dna(alignment, model=model)[1]
    }
    n <- read_tsv(file.path(dir, 'ancestralSequences.log'), col_names=c("state", "trait"), skip=3) %>%
        add_rownames() %>%
        filter(as.numeric(rowname) > (n() * burnin_pct)) %>%
        select(c(trait)) %>%
        rowwise() %>% mutate(dist=distance(trait, model, founder)) %>%
        ungroup() %>% summarize(dist=mean(dist))
    cat("\n")
    return(as.numeric(n))
}

consensus.distance <- function(dir, model='K80', founder=founder) {
    # calculate a consensus of the input sequences
    print(dir)
    x <- readDNAStringSet(file.path(dir, 'sample_aln.fa'))
    m <- consensusMatrix(x)
    mymap <- c('-'='ACGT', IUPAC_CODE_MAP)
    s <- consensusString(m, ambiguityMap=mymap, threshold=0.25)
    alignment <- pwalign(s, founder)
    dist.dna(alignment, model=model)[1]
}


prank.distance <- function(dir, prank.model='codon', model='K80', founder=founder) {
    #print(file)
    root <- prank.sequence(file.path(dir, paste0('prank_', prank.model)))
    s <- paste(root,collapse='')
    removeGaps <- function(align, gap="-") {
        myseq <- maskMotif(align, gap)
        myseq <- paste(as(myseq, "XStringViews"), collapse="")
        return(myseq)
    }
    s <- as.character(removeGaps(DNAString(s)))
    alignment <- pwalign(s, founder)
    dist.dna(alignment, model=model)[1]
}


# create a DNAbin object representing a pairwise alignment between two sequences
pwalign <- function(s1, s2) {
    ## Nucleotide global, local, and overlap alignments
    pattern <- DNAString(s1)
    subject <- DNAString(s2)

    # First use a fixed substitution matrix
    mat <- nucleotideSubstitutionMatrix(match = 1, mismatch = -3, baseOnly = FALSE)
    globalAlign <- pairwiseAlignment(pattern, subject, type="global", substitutionMatrix = mat,
                                     gapOpening = 5, gapExtension = 2)

    # extracting the aligned sequences from a pairwiseAlignment object is unncessarily complex.
    # see https://support.bioconductor.org/p/70913/#70966

    seqset <- Biostrings:::.makePostalignedSeqs(globalAlign)[1]
    return(do.call(rbind, lapply(seqset, as.DNAbin)))
}


founder <- founder.sequence()

tmp <- df 
tmp <- tmp %>% rowwise() %>% mutate(dist.pdna=prank.distance(dir, 'dna', 'K80', founder)) 
tmp <- tmp %>% rowwise() %>% mutate(dist.pcodon=prank.distance(dir, 'codon', 'K80', founder))
tmp <- tmp %>% rowwise() %>% mutate(dist.beast=beast.distance(dir, 'K80', founder))
tmp <- tmp %>% rowwise() %>% mutate(dist.consensus=consensus.distance(dir, 'K80', founder))
tmp <- tmp %>% gather(method, dist, starts_with("dist.")) %>% mutate(method=sub('dist.','',method))

tmp <- tmp %>% mutate(method=factor(method))

save(tmp,file="distances.Rda")

#pdf(file="plots.pdf")
png(file="distances.png", width=18, height=12, units = "in", res=300)

facet_labeller <- function(variable,value) {
    switch(variable,
           indel=ifelse(value, "With indels", "Without indels"),
           as.character(value))
}
ggplot(tmp, aes(x=wpi, y=dist, group=method, color=method)) +
    facet_grid(~ indel, labeller=facet_labeller) +
    geom_point(position = position_jitter(w = 0.1, h = 0.1)) +
    scale_color_discrete(name="Method",
                         breaks=c("beast", "consensus", 'pcodon', 'pdna'),
                         labels=c("Beast", "Consensus", 'Prank codon', 'Prank dna')) +
    geom_smooth(method="lm", fill=NA) 

dev.off()
# read the prank data
#	find the MRCA node

# read the founder

# Calculate distance from MRCAto founder
